README.md
# DataJoint Hub Plugin

This is an official DataJoint Python plugin for connection to DJNeuro's hosted instances. [https://djneuro.io/services](https://djneuro.io/services)

Note: This plugin adds features to the `datajoint` package and thus requires it. Users should not import directly from the plugin but instead simply `import datajoint` to access plugin features.

## Run tests locally

```
$  ./serve.sh up LOCAL
$$ docker exec -it datajoint_connection_hub-python_app_1 sh
$$ nosetests -v --tests=/home/dja/datajoint-python/tests
$$ nosetests -vw tests
$$ exit
$  ./serve.sh down LOCAL
```

Note: Make sure to have a valid Git SSH key to access necessary private repos.

CHANGELOG.md
## Release notes

### 0.0.4 -- Feb 27, 2020
* First release


tests/__init__.py
from hubapi.schema import *
from os import environ
import uuid


def setup_package():
    Org.insert([dict(org_name="datajoint", org_id=uuid.uuid4())])
    db_inst = '{}:3306'.format(environ.get('DJ_HOST', 'fakeservices.datajoint.io'))
    DatabaseInstance.insert([dict(database_dsn=db_inst, database_id=uuid.uuid4())])
    Pipeline.insert([dict(pipeline_name="travis", **(Org & 'org_name="datajoint"').fetch1('KEY'), **(DatabaseInstance & 'database_dsn="{}"'.format(db_inst)).fetch1('KEY'), pipeline_id=uuid.uuid4())])


def teardown_package():
    Pipeline.delete_quick()
    DatabaseInstance.delete_quick()
    Org.delete_quick()


tests/test_hub.py
from datajoint_connection_hub import ConnectionPlugin
import datajoint_connection_hub as hub
import datajoint as dj
from datajoint.errors import DataJointError
from nose.tools import assert_equal, raises
get_host = ConnectionPlugin.get_host


def test_normal_host():
    assert_equal(get_host('1.2.3.4'), '1.2.3.4')
    assert_equal(get_host('1.2.3.4:5678'), '1.2.3.4:5678')
    assert_equal(get_host('Ever.Green_Bear-Creek'), 'Ever.Green_Bear-Creek')
    assert_equal(get_host('Ever.Green_Bear-Creek:1234'), 'Ever.Green_Bear-Creek:1234')


def test_hub_host():
    host_input = 'hub://fakeservices.datajoint.io/datajoint/travis'
    dj.conn(host=host_input, reset=True)
    assert_equal(get_host(host_input), 'fakeservices.datajoint.io:3306')


@raises(DataJointError)
def test_hub_missing_pipeline():
    get_host('hub://fakeservices.datajoint.io/datajoint/test')


@raises(DataJointError)
def test_hub_no_tls():
    get_host('hub://fakeservices.datajoint.io:4000/datajoint/travis')


@raises(DataJointError)
def test_hub_incorrect_protocol():
    get_host('djhub://datajoint/travis')


@raises(DataJointError)
def test_hub_unreachable_server():
    get_host('hub://fakeservices.datajoint.io:4001/datajoint/travis')


@raises(DataJointError)
def test_hub_unreachable_endpoint():
    current = hub.API_TARGETS
    hub.API_TARGETS = {'PIPELINE': '/wrong_one'}
    try:
        get_host('hub://fakeservices.datajoint.io/datajoint/travis')
    except:
        hub.API_TARGETS = current
        raise


tests/test_plugin.py
from datajoint.plugin import connection_plugins


def test_check_plugin_status():
    assert(connection_plugins['hub']['verified'])


raphael_connection_hub/connection.py
import re
import requests
from requests.adapters import HTTPAdapter
from datajoint.errors import DataJointError
import pymysql as client

HUB_PROTOCOL = 'hub://'
REQUEST_PROTOCOL = 'https://'
API_ROUTE = '/api'
API_TARGETS = dict(PIPELINE='/pipeline')

session = requests.Session()
session.mount(REQUEST_PROTOCOL, HTTPAdapter(max_retries=3))


class ConnectionPlugin():
    @staticmethod
    def get_host(host_input):
        hub_path = re.findall('/[:._\-a-zA-Z0-9]+', host_input)
        if re.match(HUB_PROTOCOL, host_input) and len(hub_path) > 2:
            try:
                resp = session.get('{}{}{}{}'.format(REQUEST_PROTOCOL,
                    hub_path[0][1:], API_ROUTE, API_TARGETS['PIPELINE']),
                    params={'org_name': hub_path[1][1:], 'pipeline_name': hub_path[2][1:]},
                    timeout=10)
                if resp.status_code == 200:
                    return resp.json()[0]['database_dsn']
                elif resp.status_code == 204:
                    raise DataJointError(
                        'DataJoint Hub database resource `{}/{}/{}` not found.'.format(
                        hub_path[0][1:], hub_path[1][1:], hub_path[2][1:]))
                elif resp.status_code == 404:
                    raise DataJointError(
                        'DataJoint Hub endpoint `{}{}{}{}` unavailable.'.format(
                        REQUEST_PROTOCOL, hub_path[0][1:], API_ROUTE, API_TARGETS['PIPELINE']))
            except requests.exceptions.SSLError:
                raise DataJointError(
                    'TLS security violation on DataJoint Hub target `{}{}{}`.'.format(
                    REQUEST_PROTOCOL, hub_path[0][1:], API_ROUTE))
            except requests.exceptions.ConnectionError:
                raise DataJointError(
                    'Unable to reach DataJoint Hub target `{}{}{}`.'.format(
                    REQUEST_PROTOCOL, hub_path[0][1:], API_ROUTE))
        elif not re.match('.*/', host_input):
            return host_input
        else:
            raise DataJointError('Malformed database host `{}`.'.format(host_input))

    @staticmethod
    def connect_host(connection_obj):
        try:
            connection_obj.connect()
        except client.err.OperationalError:
            if not connection_obj.is_connected:
                target = [int(v) if i == 1 else v
                    for i, v in enumerate(
                    get_host_hook(connection_obj.conn_info['host_input']).split(':'))]
                if (target[0] != connection_obj.conn_info['host'] or len(target) > 1 and
                        target[1] != connection_obj.conn_info['port']):
                    connection_obj.conn_info['host'] = target[0]
                    connection_obj.conn_info['port'] = target[1] if len(
                        target) > 1 else connection_obj.conn_info['port']
                    connection_obj.connect()
                else:
                    raise


datajoint_connection_hub/__init__.py
import re
import requests
from requests.adapters import HTTPAdapter
from datajoint.errors import DataJointError
import pymysql as client
from .version import __version__

HUB_PROTOCOL = 'hub://'
REQUEST_PROTOCOL = 'https://'
API_ROUTE = '/api'
API_TARGETS = dict(PIPELINE='/pipeline')

session = requests.Session()
session.mount(REQUEST_PROTOCOL, HTTPAdapter(max_retries=3))


class ConnectionPlugin():
    @staticmethod
    def get_host(host_input):
        hub_path = re.findall('/[:._\-a-zA-Z0-9]+', host_input)
        if re.match(HUB_PROTOCOL, host_input) and len(hub_path) > 2:
            try:
                resp = session.get('{}{}{}{}'.format(REQUEST_PROTOCOL,
                    hub_path[0][1:], API_ROUTE, API_TARGETS['PIPELINE']),
                    params={'org_name': hub_path[1][1:], 'pipeline_name': hub_path[2][1:]},
                    timeout=10)
                if resp.status_code == 200:
                    return resp.json()[0]['database_dsn']
                elif resp.status_code == 204:
                    raise DataJointError(
                        'DataJoint Hub database resource `{}/{}/{}` not found.'.format(
                        hub_path[0][1:], hub_path[1][1:], hub_path[2][1:]))
                elif resp.status_code == 404:
                    raise DataJointError(
                        'DataJoint Hub endpoint `{}{}{}{}` unavailable.'.format(
                        REQUEST_PROTOCOL, hub_path[0][1:], API_ROUTE, API_TARGETS['PIPELINE']))
            except requests.exceptions.SSLError:
                raise DataJointError(
                    'TLS security violation on DataJoint Hub target `{}{}{}`.'.format(
                    REQUEST_PROTOCOL, hub_path[0][1:], API_ROUTE))
            except requests.exceptions.ConnectionError:
                raise DataJointError(
                    'Unable to reach DataJoint Hub target `{}{}{}`.'.format(
                    REQUEST_PROTOCOL, hub_path[0][1:], API_ROUTE))
        elif not re.match('.*/', host_input):
            return host_input
        else:
            raise DataJointError('Malformed database host `{}`.'.format(host_input))

    @staticmethod
    def connect_host(connection_obj):
        try:
            connection_obj.connect()
        except client.err.OperationalError:
            if not connection_obj.is_connected:
                target = [int(v) if i == 1 else v
                    for i, v in enumerate(
                    get_host_hook(connection_obj.conn_info['host_input']).split(':'))]
                if (target[0] != connection_obj.conn_info['host'] or len(target) > 1 and
                        target[1] != connection_obj.conn_info['port']):
                    connection_obj.conn_info['host'] = target[0]
                    connection_obj.conn_info['port'] = target[1] if len(
                        target) > 1 else connection_obj.conn_info['port']
                    connection_obj.connect()
                else:
                    raise
